UPGRADING.md

```diff
@@ -0,0 +1,372 @@
+// fmt.Println(list.TotalCount, list.Users)
+fmt.Println(list.TotalCount, list.$resource$s)
+```
+
+## Different types
+
+In the library's `v1` version, some operations would return the same type, while others wouldn't.
+
+The `v2` version moves to a more resource-oriented approach. Each API resource has a type and all types
+are defined in the `clerk` package.
+
+This means there are types for `clerk.User` and `clerk.Domain` and these types are returned by
+the Users and Domains API operations respectively.
+
+All types were consolidated and updated to match the currently supported request parameters and responses
+of the [Clerk Backend API](https://clerk.com/docs/reference/backend-api). Many fields have been renamed and a lot
+of new fields have been added. Deprecated struct fields have been dropped.
+
+### Operation parameters
+
+The `v2` version of Clerk SDK Go introduces another important change about types that can be used as API operation
+parameters. Every field for these structs is a pointer.
+
+The `v2` version of the library introduces helper functions to cast basic type values to pointers. These are:
+- `clerk.String`
+- `clerk.Bool`
+- `clerk.Int64`
+
+Using the helpers above, here's how you can invoke an API operation with a `*Params` struct.
+
+```go
+domain.Create(context.Background(), &domain.CreateParams{
+    Name: clerk.String("clerk.com"),
+    IsSatellite: clerk.Bool(true),
+})
+```
+
+You can explicitly pass zero values with `clerk.String("")` or `clerk.Int64(0)`.
+
+All API operations receive a `context.Context` as the first argument.
+
+## HTTP middleware
+
+The `v1` version of the Clerk Go SDK supports two HTTP middleware functions that can handle authentication with Clerk.
+These are `WithSessionV2` and `RequireSessionV2`.
+
+The middleware in `v1` supports authentication with a bearer token in the HTTP request headers, falling back to cookie
+based authentication if the "Authorization" header is missing.
+
+In a similar way, the `v2` version of the library also provides two middleware functions that can handle authentication
+with Clerk; `WithHeaderAuthorization` and `RequireHeaderAuthorization`.
+
+As the name implies, the new middleware support only authentication with a bearer token, that needs to be present in the
+"Authorization" header of the HTTP request.
+
+Cookie based authentication is not supported at all by the `v2` version of the library.
+
+Usage has also changed between `v1` and `v2`, as have the different options that the middleware support.
+
+Here's how you would use the `WithSessionV2` middleware to get access to the session token claims in HTTP handlers.
+
+```go
+import (
+    "net/http"
+
+    "github.com/clerkinc/clerk-sdk-go"
+)
+
+func main() {
+    client, err := clerk.NewClient("sk_live_XXX")
+    if err != nil {
+        panic(err)
+    }
+    mux := http.NewServeMux()
+    mux.Handle("/session", clerk.WithSessionV2(client)(http.HandlerFunc(handleSession)))
```

Comment:
🔧 Perhaps use the `Require*` variants in all examples? I understand verification should be the default, unless there's some other use case?

Reply:
Yes, but that seems as a design choice. Shouldn't we recommend the most secure integration?